From 0aeae91b0ca4bf0d367496d3dca7c414c724db25 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Wed, 21 Feb 2024 17:02:20 +0100
Subject: [PATCH 0654/1695] plat-stm32mp1: add function to check the supported
 OPP

Add a function to control the supported operating point
for a dedicated SoC. It will help to detect over frequency
issues.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ab8c60513965eac0a09b970cd031fd491b63bb18 plat-stm32mp1: add function to check the supported OPP]
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Change-Id: I8ebdb620d435c8a798099d79989b0c9bf627e609
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/362290
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/main.c            | 37 +++++++++++++++++++
 core/arch/arm/plat-stm32mp1/platform_config.h |  4 ++
 core/arch/arm/plat-stm32mp1/pm/context.c      |  2 -
 core/arch/arm/plat-stm32mp1/stm32_util.h      |  3 ++
 4 files changed, 44 insertions(+), 2 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index 07ecc5d0c..f70c70fa7 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -557,6 +557,43 @@ static int get_part_number(uint32_t *part_nb)
 	return 0;
 }
 
+bool stm32mp_supports_cpu_opp(uint32_t opp_id)
+{
+	uint32_t part_number = 0;
+	uint32_t id = 0;
+
+	if (get_part_number(&part_number)) {
+		DMSG("Cannot get part number\n");
+		panic();
+	}
+
+	switch (part_number) {
+#ifdef CFG_STM32MP13
+	case STM32MP135F_PART_NB:
+	case STM32MP135D_PART_NB:
+	case STM32MP133F_PART_NB:
+	case STM32MP133D_PART_NB:
+	case STM32MP131F_PART_NB:
+	case STM32MP131D_PART_NB:
+#endif
+#ifdef CFG_STM32MP15
+	case STM32MP157F_PART_NB:
+	case STM32MP157D_PART_NB:
+	case STM32MP153F_PART_NB:
+	case STM32MP153D_PART_NB:
+	case STM32MP151F_PART_NB:
+	case STM32MP151D_PART_NB:
+#endif
+		id = BIT(1);
+		break;
+	default:
+		id = BIT(0);
+		break;
+	}
+
+	return (opp_id & id) == id;
+}
+
 bool stm32mp_supports_hw_cryp(void)
 {
 	uint32_t part_number = 0;
diff --git a/core/arch/arm/plat-stm32mp1/platform_config.h b/core/arch/arm/plat-stm32mp1/platform_config.h
index c1fb658ac..920dbe218 100644
--- a/core/arch/arm/plat-stm32mp1/platform_config.h
+++ b/core/arch/arm/plat-stm32mp1/platform_config.h
@@ -369,4 +369,8 @@
 
 #define STM32MP1_CHIP_DEFAULT_VERSION	U(0)
 
+/* OPP */
+#define PLAT_MAX_OPP_NB			2U
+#define PLAT_MAX_PLLCFG_NB		6U
+
 #endif /*PLATFORM_CONFIG_H*/
diff --git a/core/arch/arm/plat-stm32mp1/pm/context.c b/core/arch/arm/plat-stm32mp1/pm/context.c
index c37256075..b9f68892e 100644
--- a/core/arch/arm/plat-stm32mp1/pm/context.c
+++ b/core/arch/arm/plat-stm32mp1/pm/context.c
@@ -35,8 +35,6 @@
 /*
  * Workaround for compilation until CPU freq OPP is integrated
  */
-#define PLAT_MAX_OPP_NB 2
-#define PLAT_MAX_PLLCFG_NB 6
 static void
 __unused stm32mp1_clk_lp_save_opp_pll1_settings(uint8_t *a __maybe_unused,
 						size_t b __maybe_unused) { };
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index b39c4f277..4140874a6 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -20,6 +20,9 @@
 TEE_Result stm32mp1_dbgmcu_get_chip_version(uint32_t *chip_version);
 TEE_Result stm32mp1_dbgmcu_get_chip_dev_id(uint32_t *chip_dev_id);
 
+/* OPP service */
+bool stm32mp_supports_cpu_opp(uint32_t opp_id);
+
 /*  Crypto HW support */
 bool stm32mp_supports_hw_cryp(void);
 
-- 
2.47.2

